// you-tube/dist/collection/components/you-tube/you-tube.js
import { Component, Prop, h } from '@stencil/core';
export class YouTube {
  componentDidLoad() {
    if ('YT' in window) {
      this.constructPlayer(window['YT']);
    }
    else {
      // load API then construct player
      this.loadApi().then(YT => this.constructPlayer(YT));
    }
  }
  // Player Constructor
  constructPlayer(YT) {
    new YT.Player('player', {
      height: this.height,
      width: this.width,
      videoId: this.video_id,
      playerVars: {
        'autoplay': this.autoplay ? parseInt(this.autoplay) : 0,
        'controls': this.controls ? parseInt(this.controls) : 1
      },
      events: {
        'onReady': event => this.onReady(event),
        'onStateChange': event => this.onStateChange(event),
        'onPlaybackQualityChange': event => this.onPlaybackQualityChange(event),
        'onApiChange': event => this.onApiChange(event),
        'onPlaybackRateChange': event => this.onPlaybackRateChange(event),
        'onError': event => this.onError(event),
      }
    });
  }
  // Event Handlers
  onReady(event) {
    if (this.playerReady) {
      this.playerReady(event);
    }
    ;
  }
  onStateChange(event) {
    if (this.playerStateChange) {
      this.playerStateChange(event);
    }
    ;
  }
  onError(event) {
    if (this.playerError) {
      this.playerError(event);
    }
    ;
  }
  onApiChange(event) {
    if (this.playerApiChange) {
      this.playerApiChange(event);
    }
    ;
  }
  onPlaybackQualityChange(event) {
    if (this.playbackQualityChange) {
      this.playbackQualityChange(event);
    }
    ;
  }
  onPlaybackRateChange(event) {
    if (this.playbackRateChange) {
      this.playbackRateChange(event);
    }
    ;
  }
  // Load the API
  loadApi() {
    const tag = document.createElement('script');
    tag.src = "https://www.youtube.com/iframe_api";
    const firstScriptTag = document.getElementsByTagName('script')[0];
    firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);
    // return a promise that resolves to YT when the iframe is ready
    return new Promise(resolve => {
      window['onYouTubeIframeAPIReady'] = () => {
        resolve(window['YT']);
      };
    });
  }
  render() {
    return h("div", { id: "player" });
  }
  static get is() { return "you-tube"; }
  static get originalStyleUrls() { return {
    "$": ["you-tube.css"]
  }; }
  static get styleUrls() { return {
    "$": ["you-tube.css"]
  }; }
  static get properties() { return {
    "width": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "width",
      "reflect": false
    },
    "height": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "height",
      "reflect": false
    },
    "video_id": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "video_id",
      "reflect": false
    },
    "autoplay": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "autoplay",
      "reflect": false
    },
    "controls": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "controls",
      "reflect": false
    },
    "playerReady": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "playerStateChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "playbackQualityChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "playbackRateChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "playerApiChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "playerError": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Function",
        "resolved": "Function",
        "references": {
          "Function": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    }
  }; }
}
